### PR TITLE
GH-2749 Include Namespaces in Statement imports from NamespaceAware collections

### PR DESCRIPTION
**Pull request by Dzeri96**
Signed-off-by: Dzerom Dzenkins <dzeri96@proton.me>

GitHub issue resolved: #2749

Briefly describe the changes proposed in this PR:

As suggested by @jeenbroekstra, I've added a call to `setNamespace` for all new namespaces coming from the imported `Statement`s, and I've removed the appropriate overrides in the SPARQL repository, as it seems like the code was almost identical. Please correct me if I'm wrong.

Since this is a WIP, there are some things to clear up:
1. Where do I put the unit tests for this? I couldn't find an appropriate place with similar tests.
2. What should we do with the deprecated `Iteration` overload of the `add` method? I suspect this might get removed with the next release so I haven't added any changes there.
3. What about the `remove` method? Should we prune unused `Namespace`s?

As soon as these questions are cleared up, and the flaky tests are fixed, I'll continue working on this PR!

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

File: core/repository/api/src/main/java/org/eclipse/rdf4j/repository/base/AbstractRepositoryConnection.java [hunk 1]
```diff
@@ -366,10 +370,23 @@ public void add(Iterable<? extends Statement> statements, Resource... contexts)
 				"contexts argument may not be null; either the value should be cast to Resource or an empty array should be supplied");
 
 		boolean localTransaction = startLocalTransaction();
+		Set<Namespace> newNamespaces = new HashSet<>();
+
 		try {
 			for (Statement st : statements) {
+				if (st instanceof NamespaceAware) {
+					newNamespaces.addAll(((NamespaceAware) st).getNamespaces());
+				}
```
Comment (jeenbroekstra) [inline core/repository/api/src/main/java/org/eclipse/rdf4j/repository/base/AbstractRepositoryConnection.java:379]: I don't think this will work - it's not `Statement` that implements `NamespaceAware`, but the collection of statements.

It's also potentially a bit of a performance burden doing the check this way. I'd suggest that instead, you do a single check outside of the loop, and then just transfer all namespaces in one go, something like this.:

```java
if (statements instanceof NamespaceAware) {
   var namespaces = ((NamespaceAware)statements).getNamespaces();
   for (namespace: namespaces) {
     setNamespace(namespace.getPrefix(), namespace.getName());
   } 
}
```

I also don't think there's need to check if the namespace already exists in the repo or not. `setNamespace` will just overwrite if it exists (which I think is the behavior we want here).

File: core/repository/sparql/src/main/java/org/eclipse/rdf4j/repository/sparql/SPARQLConnection.java [hunk 2]
```diff
@@ -642,32 +642,6 @@ public void add(Reader reader, String baseURI, RDFFormat dataFormat, Resource...
...
-			addWithoutCommit(st, contexts);
-		}
-		try {
-			conditionalCommit(localTransaction);
-		} catch (RepositoryException e) {
-			conditionalRollback(localTransaction);
-			throw e;
-		}
-	}
-
```
Comment (jeenbroekstra) [inline core/repository/sparql/src/main/java/org/eclipse/rdf4j/repository/sparql/SPARQLConnection.java:670]: I don't think you should remove this. For one thing, `SPARQLConnection` does not actually implement `setNamespace` (because there is no SPARQL command for this), so we can't support the transfer of namespace in this way when adding a Model. 

If this is something we want to support for `SPARQLConnection` as well we should probably look at ensuring that namespaces in the supplied model make it into the serialized SPARQL update that will be sent to the endpoint, somehow. This is handled in the `createInsertDataCommand` method. But to be fair even then I'm not sure you can guarantee it. There is no general standard on whether a remote SPARQL endpoint stores/reuses the the namespace definitions you supply when you do an update.

TL;DR I'd revert this change for now, and just not worry about SPARQLConnection for the time being.

**Review by jeenbroekstra — COMMENTED** [review]
Thanks for supplying this PR @Dzeri96 ! I've left a couple of suggestions for improvements. Also: can you make sure that you add some tests to cover the desired behavior? The `RepositoryConnectionTest` class (in `rdf4j-repository-testsuite`) is probably a good spot for this.

File: core/repository/api/src/main/java/org/eclipse/rdf4j/repository/base/AbstractRepositoryConnection.java (hunk 1 above)
Comment (Dzeri96) [reply core/repository/api/src/main/java/org/eclipse/rdf4j/repository/base/AbstractRepositoryConnection.java:379]: I'll get to this today, however, didn't you suggest that we skip existing namespaces in the original issue I created?

Which option should I go with?

Comment (jeenbroekstra) [reply core/repository/api/src/main/java/org/eclipse/rdf4j/repository/base/AbstractRepositoryConnection.java:379]: > I'll get to this today, however, didn't you suggest that we skip existing namespaces in the original issue I created?
> 
> Which option should I go with?

Ah good call I had forgotten that. Yes, let's go with skipping existing for now, as it's more consistent with the behavior when adding a new file.

**Comment by Dzeri96** [comment]
@jeenbroekstra Please have a look at the second iteration of my implementation. I've added a simple test, however some implementations of the `RepositoryConnectionTest` fail to execute on my machine, even on the latest `main` branch. These are:
- `ElasticSearchStoreConnectionIT`
- `HTTPStoreConnectionTest`
- `RDFSchemaRepositoryConnectionTest`
- `SPARQLStoreConnectionTest`

I've nevertheless added a skipping override for the test in the `SPARQLStoreConnectionTest`

File: core/repository/api/src/main/java/org/eclipse/rdf4j/repository/base/AbstractRepositoryConnection.java [hunk 3]
```diff
@@ -366,10 +368,22 @@ public void add(Iterable<? extends Statement> statements, Resource... contexts)
...
+
+			if (statements instanceof NamespaceAware) {
+				var newNamespaces = ((NamespaceAware) statements).getNamespaces();
+				for (Namespace newNamespace : newNamespaces) {
+					String nsPrefix = newNamespace.getPrefix();
+					if (getNamespace(nsPrefix) == null) {
+						setNamespace(nsPrefix, newNamespace.getName());
+					}
+				}
+			}
```
Comment (jeenbroekstra) [inline core/repository/api/src/main/java/org/eclipse/rdf4j/repository/base/AbstractRepositoryConnection.java:385]: Nice!

**Review by jeenbroekstra — APPROVED** [review]
Nice work @Dzeri96 - this looks good to me!

Not sure why some tests fail in your local setup, but the Github build succeeded without failure so I don't think it's your code that's the problem.
